File: C#/Assets/Scripts/Door.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Door
{
    public bool isOpen { get; set; }
    public char direction { get; set; }

    public Door(bool newIsOpen, char newDirection)
    {
        isOpen = newIsOpen;
        direction = newDirection;
    }

    public bool CanOpen(Room parentRoom, int n, int m)
    {
        bool canOpen = true;

        // Can't open an already opened door.
        if (isOpen)
        {
            canOpen = false;
        }
        // Check if we're at the end of the map.
        else if (direction == 'N' && parentRoom.coordinates[1] < 1)
        {
            canOpen = false;
        }
        else if (direction == 'E' && parentRoom.coordinates[0] > m - 2)
        {
            canOpen = false;
        }
        else if (direction == 'S' && parentRoom.coordinates[1] > n - 2)
        {
            canOpen = false;
        }
        else
        {
            if (parentRoom.coordinates[0] < 1)
            {
                canOpen = false;
            }
        }

        return canOpen;
    }
}

File: C++/RecursiveDungeonGeneration/RecursiveDungeonGeneration/Door.cpp
#include "Door.h"

Door::Door() {
	isOpen = false;
	direction = '0';
}

Door::Door(bool newIsOpen, char newDirection) {
	isOpen = newIsOpen;
	direction = newDirection;
}

void Door::setOpen(bool newIsOpen) {
    isOpen = newIsOpen;
}

bool Door::getOpen() {
    return isOpen;
}

bool Door::canOpen(int coordinates[2], int n, int m) {
	bool canOpen = true;

    if (isOpen)
    {
        canOpen = false;
    }
    else if (direction == 'N' && coordinates[1] < 1)
    {
        canOpen = false;
    }
    else if (direction == 'E' && coordinates[0] > m - 2)
    {
        canOpen = false;
    }
    else if (direction == 'S' && coordinates[1] > n - 2)
    {
        canOpen = false;
    }
    else
    {
        if (coordinates[0] < 1)
        {
            canOpen = false;
        }
    }

	return canOpen;
}
